### DISCUSSION TITLE
[General] <Title> consumed gas for the validation process in account abstraction

### DISCUSSION THREAD

**Discussion by elhajin** [zkSync-Community-Hub/zksync-developers#124] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

trying to do a lot of mathematical computation to check if the transaction is valid or not. in the **isValidSignature** function from account abstraction.

## Hello , I would greatly appreciate your assistance in clarifying a question I have 
 about account abstraction and its implications on gas calculation. During the validation step where the **bootleader** calls the **isValidSignature** function: 
-  *does the computation involved count towards the gas calculation?* 
    - If it does: 
        + what is the maximum allowable computation for this process? 
        + what happens if the validation process fails? 
*As far as I understand, the payForGas function is only called after the validation process.* : 
 - and couldn't  lead to potential issues?
  - In case of a failing transaction, couldn't  there be a risk of network vulnerability to DDoS attacks? Is there a concern that malicious actors might exploit this situation by submitting numerous transactions intentionally designed to fail but requiring substantial computational resources, potentially overloading the network's capacity?
  

### Can you share the error messages or logs you're receiving, if any?

none 

### Have you made any recent changes to the contract before encountering this issue?

none

### Are there any external libraries or contracts that your contract interacts with?

none

### Can you provide the relevant portions of your contract code where the issue is occurring?

none

### Have you tried to isolate the problem, and if so, what were the results?

none

### What steps have you already taken to try to resolve the issue?

none

**Comment by dutterbutter** [comment]
Thanks for the question! 🙏

This is a great opportunity to bring up [era-test-node](https://github.com/matter-labs/era-test-node) 🛠️. To get detailed insights into gas calculations ⛽, check out the latest [release video](https://github.com/matter-labs/era-test-node/releases/tag/v0.1.0-alpha.2). Fast-forward to the **~1:30** mark to see how it works! 

🕐 TLDR: Start `era_test_node` with the `--show-gas-details` flag 🚩 and you'll get a comprehensive breakdown on gas costs, including transaction setup, bytecode preparation, account validations, and computations.

To answer your question:

- Yes, I believe the computation in `isValidSignature` does count towards gas costs.
- For the max computation, I don't have an exact number 🤔, but I will look into it.

As for the potential risks 🚨:

- If the validation process fails, the transaction would be invalid and not proceed.
- Regarding DDoS attacks, it's less likely to be a major issue since the attacker would need to pay for the computational resources used 💸.

Hope this helps! 😊 Feel free to ask more questions! 📩

**Reply by elhajin** [reply]
Thank you for your answer @dutterbutter . I watched the video, and it's really helpful. Now, I understand that the computational resources consumed during the validation process count towards the gas cost when the transaction is valid.

**However**... I'm still confused about what happens when the validation process fails. There is no gas charged, and the computational resources consumed during this process (even if the validation fails, some computation is still required to check if the transaction is valid or not) are consumed from the era node. The node doesn't receive any gas for this work. This is why I raised the concern about potential DDoS attacks. A user could send numerous transactions in each new block that will ultimately fail, causing the node to perform a significant amount of computation without compensation. This could, at the very least, slow down the network.

I would appreciate it if you could provide further clarification on this matter.

**Reply by 4NNNN** [reply]
Hey @elhajin had the same doubt, I'm trying to implement a webauthn passkey implementation. It uses a P256 secp256r1 elliptic curve verification library, verification gas costs lesser than 300k with the gaslimit of 360k being the max. Verification doesn't work

References:
**Passkey contract** (val()--> has predefined values which works perfectly in the forge tests)- https://goerli.explorer.zksync.io/address/0x7AF6CAB20CeA9aD116212556f6B514cE18Eb8977#transactions
**verification library -** https://github.com/bcnmy/scw-contracts/blob/master/contracts/smart-account/modules/PasskeyValidationModules/Secp256r1.sol

**Reply by dutterbutter** [reply]
Sorry for the delay here. Let me investigate more on my side. I will follow up here.

**Reply by dutterbutter** [reply, marked as answer]
If validation fails, the transaction gets rejected and not included onchain (the user doesn't pay fees either). In EIP4337 validation and `payForGas` would be combined in one call, while in our case these are different methods, but are all part of the same transaction validation that has limit on computation. 

It is a general issue for account abstraction. We have several [validation rules](https://era.zksync.io/docs/reference/concepts/account-abstraction.html) to make such situation a lot harder to achieve, but more enforcements will be implemented in the future to enhance protection.

**Reply by elhajin** [reply]
thank you i got it

**Comment by biotremenag** [comment]
In the context of account abstraction, the computation involved in the isValidSignature function does count towards gas calculation. There is no explicit maximum allowable computation specified for this process, but it's essential to keep the computation within reasonable limits to ensure network efficiency and prevent abuse.

If the validation process fails, the transaction will not be included in the blockchain, and any gas spent on computation during the validation will be lost. This failure does not directly lead to a vulnerability to DDoS attacks, as the failed transaction would not consume significant network resources beyond the computation done during validation.

However, there is a concern that malicious actors could potentially submit numerous transactions intentionally designed to fail but requiring substantial computational resources. To mitigate this risk, it's crucial to carefully design and optimize the validation process to minimize unnecessary computation and ensure that failed transactions do not impose undue burden on the network. Additionally, implementing gas fees for failed transactions could disincentivize abuse and help maintain network stability.

**Reply by 4NNNN** [reply]
@elhajin @dutterbutter Are these accounts bots? Every one of my discussion has been reopened by random accounts after a months. Airdrop season has really arrived I guess lol

**Comment by davenportbri** [comment]
In the context of account abstraction, the computation involved in the isValidSignature function does count towards gas calculation. The maximum allowable computation for this process depends on the gas limit set for the transaction. If the validation process fails, the transaction is reverted, and any gas consumed during the process is lost. There is indeed a concern about potential DDoS attacks exploiting this situation by submitting numerous transactions designed to fail, consuming substantial computational resources and potentially overloading the network's capacity. Gas fees are intended to mitigate such attacks by requiring users to pay for computational resources consumed.

**Comment by loal0cksibus** [comment]
The computation involved in the isValidSignature function during the validation step does count towards gas calculation. The maximum allowable computation for this process is determined by the gas limit set for the transaction. If the validation process fails, the transaction will revert, consuming the gas provided by the sender. While payForGas is called after validation, there could potentially be concerns regarding network vulnerability to DDoS attacks if transactions requiring substantial computational resources are intentionally submitted to overload the network's capacity. It's essential to consider gas limits and transaction costs when designing smart contracts to mitigate such risks.

**Comment by moyemv** [comment]
The user is seeking clarification on whether the computation involved in the isValidSignature function during the validation step of account abstraction counts towards gas calculation, and if so, what the maximum allowable computation is. They are also concerned about the potential for DDoS attacks and network vulnerability if malicious actors intentionally submit failing transactions that require substantial computational resources. The user has not reported any error messages or logs and has not provided any relevant code or repository link.
